azure/src/main/scala/quasar/blobstore/azure/config.scala

```diff
@@ -29,14 +33,41 @@ final case class StorageUrl(value: String)
 final case class AccountName(value: String)
 final case class AccountKey(value: String)
 
-final case class AzureCredentials(accountName: AccountName, accountKey: AccountKey)
+final case class ClientId(value: String)
+final case class ClientSecret(value: String)
+final case class TenantId(value: String)
+
+sealed trait AzureCredentials extends Product with Serializable
+
+object AzureCredentials {
+  final case class SharedKey(
+    accountName: AccountName,
+    accountKey: AccountKey) extends AzureCredentials
+
+  final case class ActiveDirectory(
+    clientId: ClientId,
+    tenantId: TenantId,
+    clientSecret: ClientSecret) extends AzureCredentials
+}
 
 final case class MaxQueueSize(value: Int Refined Positive)
 
 object MaxQueueSize {
   def default: MaxQueueSize = MaxQueueSize(10)
 }
 
+final case class Expires[A](value: A, expiresAt: OffsetDateTime)
```

Comment:
Opinions on this? This is just a newtyped `(,) a` and its corresponding `Functor` instance.

Reply:
Looks good to me!